src/domain/prices/prices.repository.ts

```diff
@@ -0,0 +1,63 @@
+import { Inject, Injectable } from '@nestjs/common';
+import {
+  ILoggingService,
+  LoggingService,
+} from '../../logging/logging.interface';
+import { IPricesApi } from '../interfaces/prices-api.interface';
+import { IPricesRepository } from './prices.repository.interface';
+import { isArray } from 'lodash';
+
+@Injectable()
+export class PricesRepository implements IPricesRepository {
+  constructor(
+    @Inject(LoggingService) private readonly loggingService: ILoggingService,
+    @Inject(IPricesApi) private readonly pricesApi: IPricesApi,
+  ) {}
+
+  async getNativeCoinPrice(args: {
+    nativeCoinId: string;
+    fiatCode: string;
+  }): Promise<number> {
+    try {
+      const result = await this.pricesApi.getNativeCoinPrice(args);
+      const { nativeCoinId, fiatCode } = args;
+      const nativeCoinPrice = result?.[nativeCoinId]?.[fiatCode];
+      if (!nativeCoinPrice) {
+        this.loggingService.error(`Got an invalid ${nativeCoinId} price`);
```

Comment:
This error does not seem correct. Would our Fiat Price provider return an "invalid" price for a token/native coin if it exists?

Or is it more that the token/native coin does not exist, or the conversion to that specific code does not exist?

Reply:
It could be both cases. The provider is not helping here because the status codes and the payloads are not very helpful. The status code is always 200, no matter the kind of absence (token doesn't exist, or fiat code is not supported). Examples:
- Token doesn't exist (made up address): https://api.coingecko.com/api/v3/simple/token_price/ethereum?contract_addresses=0xa0b86991c6218b36c1d19d4b2e9eb0ce3606eb48&vs_currencies=eur
- Native coin doesn't exist: https://api.coingecko.com/api/v3/simple/price?ids=foo&vs_currencies=eur
- Token exists but fiat code is not supported: https://api.coingecko.com/api/v3/simple/token_price/ethereum?contract_addresses=0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48&vs_currencies=foo

So for the sake of simplicity, I'm just inspecting the response and throwing an error if it doesn't contain the price. Of course, we could implement some logic to distinguish between those scenarios and throw ad-hoc errors. That could make the code a bit more complex but I'd be happy to add this if you think it's valuable.

Reply:
In that case and to start simple I would log the conditions on which this error happened. You can use a structure to help you in this case.

Something like:
```typescript
this.loggingService.error({
  type: 'invalid_native_coin_price',
  native_coin_id: nativeCoinId,
  fiat_code: fiatCode,
});
```

You can do the same for `getTokenPrice`

Also probably to use the `warn` log level here as an indication that something unexpected happened (as you have it in `getTokenPrice`). Or did you want to distinguish these two?

Reply:
My initial attempt was to distinguish between token price absence and native coin price absence and assign a `warn` level to the former and an `error` level to the latter, which seems more critical. But let's start simple as you said, and assign a `warn` level to both.

I'm going to set prices to`null` for each item whose price cannot be retrieved, so the routes layer can map to `0` to fulfill the contract with the clients.

Reply:
Changed in [17040f3](https://github.com/safe-global/safe-client-gateway/pull/732/commits/17040f3f07ad63a3682ce6b665dc451dfae545a2)